subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DefaultPendingDependenciesHandler.java

```diff
@@ -36,11 +37,20 @@ public Visitor start() {
         private List<PendingDependencies> noLongerPending;
 
         public boolean maybeAddAsPendingDependency(NodeState node, DependencyState dependencyState) {
-            ModuleIdentifier key = dependencyState.getModuleIdentifier();
             boolean isOptionalDependency = dependencyState.getDependency().isPending();
+            ComponentSelector selector = dependencyState.getDependency().getSelector();
+            String group;
+            String module;
+            if (selector instanceof ModuleComponentSelector) {
+                group = ((ModuleComponentSelector) selector).getGroup();
+                module = ((ModuleComponentSelector) selector).getModule();
+            } else {
+                return false;
+            }
```

Comment:
I suspect that we'd get the same performance improvement by first returning `false` for anything other than `ModuleComponentSelector`, and then simply using `dependencyState.getModuleIdentifier()`. We already have a cache for `group + module -> moduleIdentifier` so this should be used (and we could pre-compute the hashcode).

My guess is that the expensive part of  `dependencyState.getModuleIdentifier()` is due to project components. If this assumption is wrong, then there's something wrong with `ImmutableModuleIdentifierFactory`.

:x: I'd like to verify before merging this, to avoid adding this additional complexity if it isn't required.

To be clear, I'm suggesting adding this to the top of the existing method:
```
if (!(depenendencyState.getDependency().getSelector() instanceof ModuleComponentSelector)) {
    return false;
}
```

Reply:
I'd be *very* surprised if it was because of project identifiers. It clearly shows as a something costly in the visit outgoing dependencies call:

![selection_132](https://user-images.githubusercontent.com/316357/35848323-00a3deb0-0b1e-11e8-9f8c-9cd325993015.png)

To me the problem is not about the factory, but the fact we have a large number of identifiers. It's really hurting us that we have to create those. However I'm wondering if doing this optimization is not just deferring the call to `getModuleIdentifier` to later, at least for dependencies which *are* going to be used.

Reply:
⭕️   Looking at the flame graphs,`getModuleIdentifier` is also called for `NodeState.isExcluded` and much more often than for pending dependencies. So you're still paying most of the cost. I don' think we actually need a module identifier for exclusions, since they are completely controlled by us.

Reply:
That's correct. In fact, we should avoid using `ModuleIdentifier` altogether, but that's not an easy move.

Reply:
I don't really understand how `DefaultModuleIdentifier` can be so expensive. It's just a simple class with 2 final fields. What makes it expensive to construct? We seem to be blaming this simple class for a big performance impact, and it feels wrong to me.

With `DefaultImmutableModuleIdentifierFactory` we already have a cache of instances. Is it maybe thread synchronization that's causing the issues, because we have `ConcurrentMap<String, ConcurrentMap<String, ModuleIdentifier>>`?

I think it would be worth trying to change `DefaultImmutableModuleIdentifierFactory.module()` to simply create a new instance to see if that removes the hot spot.

Reply:
Ok let's step back a bit. There are different things:

1. we introduced the factory to deduplicate module identifiers, because they are in practice very redundant and they used to cause excessive memory usage and garbage collection
2. deduplication is done by doing some "simple" hashmap lookups
3. the process is already efficient

The problem is that we create a _lot_ of those identifiers. So even if each lookup is cheap, it turns out that we create a crazy amount of those identifiers. Any way we can reduce the number of identifiers we create is better. Because hashmaps, despite fast, don't prevent a call to `.hashCode()` and `.equals()` on strings. The longer the identifier is, the slower it is.

Reply:
OK I believe you, but something still smells off to me. We create lots of other small objects (like `ModuleComponentIdentifier`) and we haven't tried to cache those. I thought the JVM was very good as creating and cleaning up these small, short-lived types.

I guess my gut-feel reaction is wrong.

Reply:
I don't think your gut feeling is bad @bigdaz . The caching lead to calling equals/hashCode on those Strings a lot more, which is the CPU overhead we are now struggling with. De-duplication is only necessary for objects that are kept in memory for a long time (e.g. in the ArtifactCollection of a Configuration). If that's the problem, we could do the de-duplication only when we build the result, but not inside the graph builder where the objects are mostly short-lived. Short-lived objects are collected quickly in the young generation.

Reply:
That's an interesting approach to try. However if I recall when this was introduced, it was precisely because the memory used by the graph became huge (they are only short lived in the sense that they are discarded after resolution, but we keep them for the lifetime of resolution).

Other possible (but way trickier for correctness) approaches would be to conflate "similar" objects so that they implement the `ModuleIdentifier` interface (for example, selectors have the very same information, plus the version). It's trickier because `equals` wouldn't do what we think, when we need an `equals` for an identifier.

Reply:
If I had to investigate this, I'd use JFR to see GC statistics and allocation traces with/without this cache. Might be worth doing to get a clear picture of where this is helping and where it is hurting.